homeassistant/components/vacuum/xiaomi.py

```diff
@@ -0,0 +1,373 @@
+    SERVICE_STOP_REMOTE_CONTROL: {'method': 'async_stop_remote_control'},
+    SERVICE_MOVE_REMOTE_CONTROL: {
+        'method': 'async_move_remote_control',
+        'schema': SERVICE_SCHEMA_REMOTE_CONTROL},
+    SERVICE_MOVE_REMOTE_CONTROL_STEP: {
+        'method': 'async_move_remote_control_step',
+        'schema': SERVICE_SCHEMA_REMOTE_CONTROL},
+}
+
+SUPPORT_XIAOMI = SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_PAUSE | \
+                 SUPPORT_STOP | SUPPORT_RETURN_HOME | SUPPORT_FANSPEED | \
+                 SUPPORT_SENDCOMMAND | SUPPORT_LOCATE | \
+                 SUPPORT_STATUS | SUPPORT_BATTERY
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Xiaomi vacuum cleaner robot platform."""
+    if PLATFORM not in hass.data:
+        hass.data[PLATFORM] = {}
+
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+    token = config.get(CONF_TOKEN)
+
+    # Create handler
+    mirobo = MiroboVacuum(hass, name, host, token)
+    hass.data[PLATFORM][host] = mirobo
+
+    async_add_devices([mirobo], update_before_add=True)
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Map services to methods on MiroboVacuum."""
+        _LOGGER.debug('XIAOMI async_service_handler -> %s, %s',
+                      service.service, service.data)
+        method = SERVICE_TO_METHOD.get(service.service)
+        if not method:
+            return
+
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids:
+            target_vacuums = [vac for vac in hass.data[PLATFORM].values()
+                              if vac.entity_id in entity_ids]
+        else:
+            target_vacuums = hass.data[PLATFORM].values()
+
+        update_tasks = []
+        for vacuum in target_vacuums:
+            yield from getattr(vacuum, method['method'])(**params)
+
+        for vacuum in target_vacuums:
+            if vacuum.should_poll:
+                update_coro = vacuum.async_update_ha_state(True)
+                update_tasks.append(update_coro)
+
+        if update_tasks:
+            yield from asyncio.wait(update_tasks, loop=hass.loop)
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml'))
+
+    for vacuum_service in SERVICE_TO_METHOD:
+        schema = SERVICE_TO_METHOD[vacuum_service].get(
+            'schema', VACUUM_SERVICE_SCHEMA)
+        hass.services.async_register(
+            DOMAIN, vacuum_service, async_service_handler,
+            description=descriptions.get(vacuum_service), schema=schema)
+
+
+class MiroboVacuum(VacuumDevice):
+    """Representation of a Xiaomi Vacuum cleaner robot."""
+
+    def __init__(self, hass, name, host, token):
```

Comment:
oops, I think that is there from when the class was a simple handler, fixing it!